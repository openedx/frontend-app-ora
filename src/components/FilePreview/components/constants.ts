import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from './FileRenderer/BaseRenderers';

import messages from './messages';

export const errorStatuses = {
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
} as const;

export const fileTypes = {
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
} as const;

export const errorMessages = {
  [errorStatuses.notFound]: messages.fileNotFoundError,
  [errorStatuses.serverError]: messages.unknownError,
};

/**
 * Config data
 */
export const renderers = {
  [fileTypes.pdf]: PDFRenderer,
  [fileTypes.jpg]: ImageRenderer,
  [fileTypes.jpeg]: ImageRenderer,
  [fileTypes.bmp]: ImageRenderer,
  [fileTypes.png]: ImageRenderer,
  [fileTypes.txt]: TXTRenderer,
  [fileTypes.gif]: ImageRenderer,
  [fileTypes.jfif]: ImageRenderer,
  [fileTypes.pjpeg]: ImageRenderer,
  [fileTypes.pjp]: ImageRenderer,
  [fileTypes.svg]: ImageRenderer,
};

export const supportedTypes = Object.keys(renderers);
