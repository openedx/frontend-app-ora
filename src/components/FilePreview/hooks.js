import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';

import messages from './messages';

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
});

/**
 * Config data
 */
export const RENDERERS = {
  [FileTypes.pdf]: PDFRenderer,
  [FileTypes.jpg]: ImageRenderer,
  [FileTypes.jpeg]: ImageRenderer,
  [FileTypes.bmp]: ImageRenderer,
  [FileTypes.png]: ImageRenderer,
  [FileTypes.txt]: TXTRenderer,
  [FileTypes.gif]: ImageRenderer,
  [FileTypes.jfif]: ImageRenderer,
  [FileTypes.pjpeg]: ImageRenderer,
  [FileTypes.pjp]: ImageRenderer,
  [FileTypes.svg]: ImageRenderer,
};

export const SUPPORTED_TYPES = Object.keys(RENDERERS);

export const ERROR_STATUSES = {
  [ErrorStatuses.notFound]: messages.fileNotFoundError,
  [ErrorStatuses.serverError]: messages.unknownError,
};

/**
 * Util methods and transforms
 */
export const getFileType = (fileName) => fileName.split('.').pop()?.toLowerCase();
export const isSupported = (file) => SUPPORTED_TYPES.includes(
  getFileType(file.name),
);

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const renderHooks = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    ERROR_STATUSES[errorStatus] || ERROR_STATUSES[ErrorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = RENDERERS[getFileType(file.name)];
  const rendererProps = {
    fileName: file.name,
    url: file.downloadUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
