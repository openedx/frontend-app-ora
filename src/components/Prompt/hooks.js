import { useState } from 'react';

const usePromptHooks = () => {
  const [open, setOpen] = useState(true);

  const toggleOpen = () => setOpen(!open);

  return {
    open,
    toggleOpen,
  };
};

export default usePromptHooks;
