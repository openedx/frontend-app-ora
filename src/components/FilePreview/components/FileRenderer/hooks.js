import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
