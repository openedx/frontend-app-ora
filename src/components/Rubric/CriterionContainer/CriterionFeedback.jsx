import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { feedbackRequirement } from 'data/services/lms/constants';

import messages from './messages';

/**
 * <CriterionFeedback />
 */
const CriterionFeedback = ({ criterion }) => {
  const { formatMessage } = useIntl();

  let commentMessage = formatMessage(messages.addComments);
  if (criterion.feedbackRequired === feedbackRequirement.optional) {
    commentMessage += ` ${formatMessage(messages.optional)}`;
  }

  const { feedbackValue, feedbackIsInvalid, feedbackOnChange } = criterion;

  if (
    !criterion.feedbackEnabled
    || criterion.feedbackRequired === feedbackRequirement.disabled
  ) {
    return null;
  }

  return (
    <Form.Group isInvalid={feedbackIsInvalid}>
      <Form.Control
        as="textarea"
        className="criterion-feedback feedback-input"
        floatingLabel={commentMessage}
        value={feedbackValue}
        onChange={feedbackOnChange}
      />
      {feedbackIsInvalid && (
        <Form.Control.Feedback type="invalid" className="feedback-error-msg">
          {formatMessage(messages.criterionFeedbackError)}
        </Form.Control.Feedback>
      )}
    </Form.Group>
  );
};

CriterionFeedback.propTypes = {
  criterion: PropTypes.shape({
    feedbackValue: PropTypes.string.isRequired,
    feedbackIsInvalid: PropTypes.bool.isRequired,
    feedbackOnChange: PropTypes.func.isRequired,
    feedbackEnabled: PropTypes.bool.isRequired,
    feedbackRequired: PropTypes.oneOf(Object.values(feedbackRequirement)).isRequired,
  }).isRequired,
};

export default CriterionFeedback;
