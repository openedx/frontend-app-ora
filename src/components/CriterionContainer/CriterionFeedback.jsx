import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { useViewStep } from 'hooks/routing';
import { stepNames } from 'constants';
import {
  useShowValidation,
  useCriterionFeedbackFormFields,
} from 'hooks/assessment';

import messages from './messages';

/**
 * <CriterionFeedback />
 */
const CriterionFeedback = ({ criterion, criterionIndex }) => {
  console.log({ CriterionFeedback: { criterion, criterionIndex } });
  const { formatMessage } = useIntl();
  const step = useViewStep();
  const showValidation = useShowValidation();
  const { isInvalid, value, onChange } = useCriterionFeedbackFormFields(criterionIndex);

  if (step === stepNames.studentTraining) {
    return null;
  }

  const { feedbackEnabled, feedbackRequired } = criterion;
  if (!feedbackEnabled) {
    return null;
  }
  let commentMessage = formatMessage(messages.addComments);
  if (!feedbackRequired) {
    commentMessage += ` ${formatMessage(messages.optional)}`;
  }

  return (
    <Form.Group isInvalid={showValidation && isInvalid}>
      <Form.Control
        as="textarea"
        className="criterion-feedback feedback-input"
        floatingLabel={commentMessage}
        value={value}
        onChange={onChange}
      />
      {showValidation && isInvalid && (
        <Form.Control.Feedback type="invalid" className="feedback-error-msg">
          {formatMessage(messages.criterionFeedbackError)}
        </Form.Control.Feedback>
      )}
    </Form.Group>
  );
};

CriterionFeedback.propTypes = {
  criterion: PropTypes.shape({
    feedbackEnabled: PropTypes.bool.isRequired,
    feedbackRequired: PropTypes.bool.isRequired,
  }).isRequired,
  criterionIndex: PropTypes.number.isRequired,
};

export default CriterionFeedback;
