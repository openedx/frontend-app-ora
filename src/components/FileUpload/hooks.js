import { useCallback } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import { useDownloadFiles } from 'hooks/app';

export const stateKeys = StrictDict({
  shouldShowError: 'shouldShowError',
  isModalOpen: 'isModalOpen',
  uploadArgs: 'uploadArgs',
  description: 'description',
});

export const useUploadConfirmModalHooks = ({
  file,
  closeHandler,
  uploadHandler,
}) => {
  const [description, setDescription] = useKeyedState(stateKeys.description, '');
  const [shouldShowError, setShouldShowError] = useKeyedState(
    stateKeys.shouldShowError,
    false,
  );

  const confirmUploadClickHandler = () => {
    // console.log({ confirmUploadClick: { description } });
    if (description !== '') {
      uploadHandler(file, description);
    } else {
      setShouldShowError(true);
    }
  };

  const exitHandler = () => {
    setShouldShowError(false);
    setDescription('');
    closeHandler();
  };

  // Modifying pointer of file object. This is not a good practice.
  // eslint-disable-next-line no-param-reassign, no-return-assign
  const onFileDescriptionChange = (event) => setDescription(event.target.value);

  return {
    shouldShowError,
    confirmUploadClickHandler,
    exitHandler,
    onFileDescriptionChange,
  };
};

export const useFileUploadHooks = ({ onFileUploaded }) => {
  const [uploadArgs, setUploadArgs] = useKeyedState(stateKeys.uploadArgs, {});
  const [isModalOpen, setIsModalOpen] = useKeyedState(
    stateKeys.isModalOpen,
    false,
  );

  const confirmUpload = useCallback(async (file, description) => {
    // console.log({ confirmUpload: { file, description } });
    setIsModalOpen(false);
    if (onFileUploaded) {
      // console.log({ uploadArgs });
      await onFileUploaded({ ...uploadArgs, description });
    }
    setUploadArgs({});
  }, [uploadArgs, onFileUploaded, setIsModalOpen, setUploadArgs]);

  const closeUploadModal = useCallback(() => {
    setIsModalOpen(false);
    setUploadArgs({});
  }, [setIsModalOpen, setUploadArgs]);

  const onProcessUpload = useCallback(
    ({ fileData, handleError, requestConfig }) => {
      setIsModalOpen(true);
      setUploadArgs({ fileData, handleError, requestConfig });
    },
    [setIsModalOpen, setUploadArgs],
  );

  return {
    isModalOpen,
    uploadArgs,
    confirmUpload,
    closeUploadModal,
    onProcessUpload,
  };
};

export const useFileDownloadHooks = ({ files, zipFileName }) => {
  const downloadFileMation = useDownloadFiles();
  const downloadFiles = () => downloadFileMation.mutate({
    files,
    zipFileName,
  });
  return {
    downloadFiles,
    status: downloadFileMation.status,
  };
};

export default {
  useUploadConfirmModalHooks,
  useFileUploadHooks,
  useFileDownloadHooks,
};
