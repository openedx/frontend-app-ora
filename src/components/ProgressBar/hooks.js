import { useParams } from 'react-router-dom';
import { useIsEmbedded, useViewStep } from 'hooks';
import {
  useEffectiveGrade,
  useGlobalState,
} from 'data/services/lms/hooks/selectors';
import {
  stepRoutes,
  stepStates,
} from 'data/services/lms/constants';

export const useProgressStepData = ({ step, canRevisit = false }) => {
  const { xblockId, courseId } = useParams();
  const isEmbedded = useIsEmbedded();
  const viewStep = useViewStep();
  const { stepState } = useGlobalState({ step });

  const href = `/${stepRoutes[step]}${isEmbedded ? '/embedded' : ''}/${courseId}/${xblockId}`;
  const isActive = viewStep === step;
  const isEnabled = (
    isActive
    || (stepState === stepStates.inProgress)
    || (canRevisit && stepState === stepStates.done)
  );
  const myGrade = useEffectiveGrade()?.stepScore;
  return {
    href,
    isEnabled,
    isActive,
    isComplete: stepState === stepStates.done,
    inProgress: stepState === stepStates.inProgress,
    isPastDue: stepState === stepStates.closed,
    myGrade,
    // myGrade: { earned: 8, possible: 10 },
    // isPastDue: step === 'self',
  };
};

export default { useProgressStepData };
