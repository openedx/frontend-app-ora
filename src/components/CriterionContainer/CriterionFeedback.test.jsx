import { shallow } from '@edx/react-unit-test-utils';

import { useViewStep } from 'hooks/routing';
import { stepNames } from 'constants/index';
import {
  useShowValidation,
  useCriterionFeedbackFormFields,
} from 'hooks/assessment';

import CriterionFeedback from './CriterionFeedback';

jest.mock('hooks/assessment');
jest.mock('hooks/routing');

describe('<CriterionFeedback />', () => {
  const props = {
    criterion: {
      feedbackEnabled: true,
      feedbackRequired: true,
    },
    criterionIndex: 0,
  };

  it('render empty on student training', () => {
    useViewStep.mockReturnValue(stepNames.studentTraining);
    useCriterionFeedbackFormFields.mockReturnValue({
      value: '',
      onChange: jest.fn(),
      isInvalid: false,
    });
    const wrapper = shallow(<CriterionFeedback {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('render empty on feedback not enable', () => {
    useViewStep.mockReturnValue(stepNames.self);
    useCriterionFeedbackFormFields.mockReturnValue({
      value: '',
      onChange: jest.fn(),
      isInvalid: false,
    });

    const wrapper = shallow(<CriterionFeedback
      {...props}
      criterion={{
        feedbackEnabled: false,
        feedbackRequired: false,
      }}
    />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('render with validation error', () => {
    useViewStep.mockReturnValue(stepNames.self);
    useShowValidation.mockReturnValue(true);
    useCriterionFeedbackFormFields.mockReturnValue({
      value: '',
      onChange: jest.fn(),
      isInvalid: true,
    });

    const wrapper = shallow(<CriterionFeedback {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('Form.Control.Feedback').length).toBe(1);
  });

  it('render without validation error', () => {
    useViewStep.mockReturnValue(stepNames.self);
    useShowValidation.mockReturnValue(false);
    useCriterionFeedbackFormFields.mockReturnValue({
      value: '',
      onChange: jest.fn(),
      isInvalid: false,
    });

    const wrapper = shallow(<CriterionFeedback {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.instance.findByType('Form.Control.Feedback').length).toBe(0);
  });
});
