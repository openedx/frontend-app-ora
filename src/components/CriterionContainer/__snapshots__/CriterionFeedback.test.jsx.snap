// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CriterionFeedback /> render empty on feedback not enable 1`] = `null`;

exports[`<CriterionFeedback /> render empty on student training 1`] = `null`;

exports[`<CriterionFeedback /> render with validation error 1`] = `
<Form.Group
  isInvalid={true}
>
  <Form.Control
    as="textarea"
    className="criterion-feedback feedback-input"
    floatingLabel="Add comments"
    onChange={[MockFunction]}
    value=""
  />
  <Form.Control.Feedback
    className="feedback-error-msg"
    type="invalid"
  >
    The feedback is required
  </Form.Control.Feedback>
</Form.Group>
`;

exports[`<CriterionFeedback /> render without validation error 1`] = `
<Form.Group
  isInvalid={false}
>
  <Form.Control
    as="textarea"
    className="criterion-feedback feedback-input"
    floatingLabel="Add comments"
    onChange={[MockFunction]}
    value=""
  />
</Form.Group>
`;
